testdata/p4_16_samples/v1model-special-ops-bmv2.p4

```diff
@@ -0,0 +1,470 @@
+                          in bit<8> byte3)    // least significant byte
+{
+    apply {
+        ipv4_address = (((bit<32>) byte0 << 24) |
+            ((bit<32>) byte1 << 16) |
+            ((bit<32>) byte2 << 8) |
+            (bit<32>) byte3);
+    }
+}
+
+
+control ingress(inout headers_t hdr,
+                inout meta_t meta,
+                inout standard_metadata_t standard_metadata)
+{
+#ifdef ENABLE_DEBUG_TABLES
+    debug_std_meta() debug_std_meta_ingress_start;
+    debug_std_meta() debug_std_meta_ingress_end;
+    my_debug_1() my_debug_1_1;
+    my_debug_1() my_debug_1_2;
+#endif  // ENABLE_DEBUG_TABLES
+    fill_ipv4_address() c_fill_ipv4_address;
+
+    const bit<32> RESUBMITTED_PKT_L2PTR = 0xe50b;
+    const bit<32> RECIRCULATED_PKT_L2PTR = 0xec1c;
+
+    action set_l2ptr(bit<32> l2ptr) {
+        meta.fwd.l2ptr = l2ptr;
+    }
+    action set_mcast_grp(bit<16> mcast_grp) {
+        standard_metadata.mcast_grp = mcast_grp;
+    }
+    action do_resubmit(bit<32> new_ipv4_dstAddr) {
+        hdr.ipv4.dstAddr = new_ipv4_dstAddr;
+        // By giving a list of fields inside the curly braces { } to
+        // resubmit, when things go well p4c creates a field list of
+        // those field names in the BMv2 JSON file output by the
+        // compiler.  All of those field names should have their
+        // values preserved from the packet being processed now, to
+        // the packet that will be processed by the ingress control
+        // block in the future.
+
+        // Note: There is what might be considered a bug in p4c that
+        // if you give an individual metadata field name in the list
+        // below, and the compiler can simplify it to a constant
+        // value, e.g. because you assign that field a constant value
+        // shortly before the resubmit() call, then the BMv2 JSON file
+        // will have that constant value in it instead of the field
+        // name.  I believe in that case that BMv2 simple_switch will
+        // _not_ have that metadata field value preserved.
+
+        // If you give an entire struct like standard_metadata, it
+        // includes all fields inside of that struct.  Even though
+        // standard_metadata.instance_type is one of the fields inside
+        // of this struct, that field's value will _not_ be preserved
+        // across the resubmit operation -- the resubmitted packet
+        // will have standard_metadata.instance_type ==
+        // BMV2_V1MODEL_INSTANCE_TYPE_RESUBMIT.  The instance_type
+        // field is an exception to the "preserve metadata field
+        // value" rule.
+
+        // For the resubmit operation, standard_metadata.resubmit_flag
+        // is 0 for the resubmitted packet, which avoids that packet
+        // being resubmitted indefinitely, unless your program
+        // explicitly causes a separate call to resubmit() on each
+        // execution of ingress.
+        resubmit(standard_metadata);
```

Comment:
You should an explicit type argument to resubmit if you want field names.
`resubmit<standard_metadata_t>(standard_metadata)`

Reply:
I tried making that change to the source code, and at least with the version of p4c I compiled it with, it made no difference in the resulting compiled BMv2 JSON file.  It contained the list of field names with or without the `<standard_metadata_t>`.

Reply:
I believe at least one of the intermediate P4_16 programs generated by the compiler (maybe after front end, maybe after mid end) has this `<standard_metadata_t>` type specifier added to the P4_16 code automatically, so it shouldn't depend upon the back end.

Reply:
that makes a difference when you use a constant (list) as an argument.